File: CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;

namespace SwapRows
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
		private DevExpress.XtraGrid.GridControl gridControl1;
		private DevExpress.XtraGrid.Views.Grid.GridView gridView1;
		private System.Data.DataSet dataSet1;
		private System.Data.DataTable dataTable1;
		private System.Data.DataColumn dataColumn1;
		private System.Data.DataColumn dataColumn2;
		private System.Data.DataColumn dataColumn3;
		private System.Windows.Forms.Button button1;
		private System.Windows.Forms.Button button2;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent() {
			this.gridControl1 = new DevExpress.XtraGrid.GridControl();
			this.gridView1 = new DevExpress.XtraGrid.Views.Grid.GridView();
			this.dataSet1 = new System.Data.DataSet();
			this.dataTable1 = new System.Data.DataTable();
			this.dataColumn1 = new System.Data.DataColumn();
			this.dataColumn2 = new System.Data.DataColumn();
			this.dataColumn3 = new System.Data.DataColumn();
			this.button1 = new System.Windows.Forms.Button();
			this.button2 = new System.Windows.Forms.Button();
			((System.ComponentModel.ISupportInitialize)(this.gridControl1)).BeginInit();
			((System.ComponentModel.ISupportInitialize)(this.gridView1)).BeginInit();
			((System.ComponentModel.ISupportInitialize)(this.dataSet1)).BeginInit();
			((System.ComponentModel.ISupportInitialize)(this.dataTable1)).BeginInit();
			this.SuspendLayout();
			// 
			// gridControl1
			// 
			this.gridControl1.Dock = System.Windows.Forms.DockStyle.Left;
			// 
			// gridControl1.EmbeddedNavigator
			// 
			this.gridControl1.EmbeddedNavigator.Name = "";
			this.gridControl1.Location = new System.Drawing.Point(0, 0);
			this.gridControl1.MainView = this.gridView1;
			this.gridControl1.Name = "gridControl1";
			this.gridControl1.Size = new System.Drawing.Size(267, 235);
			this.gridControl1.TabIndex = 0;
			this.gridControl1.ViewCollection.AddRange(new DevExpress.XtraGrid.Views.Base.BaseView[] {
																										this.gridView1});
			// 
			// gridView1
			// 
			this.gridView1.GridControl = this.gridControl1;
			this.gridView1.Name = "gridView1";
			// 
			// dataSet1
			// 
			this.dataSet1.DataSetName = "NewDataSet";
			this.dataSet1.Locale = new System.Globalization.CultureInfo("en-US");
			this.dataSet1.Tables.AddRange(new System.Data.DataTable[] {
																		  this.dataTable1});
			// 
			// dataTable1
			// 
			this.dataTable1.Columns.AddRange(new System.Data.DataColumn[] {
																			  this.dataColumn1,
																			  this.dataColumn2,
																			  this.dataColumn3});
			this.dataTable1.TableName = "Table1";
			// 
			// dataColumn1
			// 
			this.dataColumn1.ColumnName = "ID";
			this.dataColumn1.DataType = typeof(int);
			// 
			// dataColumn2
			// 
			this.dataColumn2.ColumnName = "Keyword";
			// 
			// dataColumn3
			// 
			this.dataColumn3.ColumnName = "Order";
			this.dataColumn3.DataType = typeof(int);
			// 
			// button1
			// 
			this.button1.Location = new System.Drawing.Point(280, 28);
			this.button1.Name = "button1";
			this.button1.Size = new System.Drawing.Size(62, 20);
			this.button1.TabIndex = 1;
			this.button1.Text = "&Up";
			this.button1.Click += new System.EventHandler(this.button1_Click);
			// 
			// button2
			// 
			this.button2.Location = new System.Drawing.Point(280, 62);
			this.button2.Name = "button2";
			this.button2.Size = new System.Drawing.Size(62, 20);
			this.button2.TabIndex = 2;
			this.button2.Text = "&Down";
			this.button2.Click += new System.EventHandler(this.button2_Click);
			// 
			// Form1
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(353, 235);
			this.Controls.Add(this.button2);
			this.Controls.Add(this.button1);
			this.Controls.Add(this.gridControl1);
			this.Name = "Form1";
			this.Text = "Form1";
			this.Load += new System.EventHandler(this.Form1_Load);
			((System.ComponentModel.ISupportInitialize)(this.gridControl1)).EndInit();
			((System.ComponentModel.ISupportInitialize)(this.gridView1)).EndInit();
			((System.ComponentModel.ISupportInitialize)(this.dataSet1)).EndInit();
			((System.ComponentModel.ISupportInitialize)(this.dataTable1)).EndInit();
			this.ResumeLayout(false);

		}
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

		void FillDataSet() {
			dataTable1.Rows.Add(new object[] {1, "abstract", 1});
			dataTable1.Rows.Add(new object[] {2, "event", 2});
			dataTable1.Rows.Add(new object[] {3, "new", 3});
			dataTable1.Rows.Add(new object[] {4, "struct", 4});
			dataTable1.Rows.Add(new object[] {5, "as", 5});
			dataTable1.Rows.Add(new object[] {6, "explicit", 6});
			dataTable1.Rows.Add(new object[] {7, "null", 7});
			dataTable1.Rows.Add(new object[] {8, "switch", 8});
			dataTable1.Rows.Add(new object[] {9, "base", 9});
			dataTable1.Rows.Add(new object[] {10, "extern", 10});
			dataTable1.Rows.Add(new object[] {11, "object", 11});
			dataTable1.Rows.Add(new object[] {12, "this", 12});
			dataTable1.Rows.Add(new object[] {13, "bool", 13});
			dataTable1.Rows.Add(new object[] {14, "false", 14});
			dataTable1.Rows.Add(new object[] {15, "operator", 15});
			dataTable1.Rows.Add(new object[] {16, "throw", 16});
			dataTable1.Rows.Add(new object[] {17, "break", 17});
			dataTable1.AcceptChanges();
		}

		private void Form1_Load(object sender, System.EventArgs e) {
			FillDataSet();
			gridControl1.DataSource = dataTable1;
			gridView1.PopulateColumns();
			gridView1.Columns[OrderFieldName].SortOrder = DevExpress.Data.ColumnSortOrder.Ascending;
			gridView1.OptionsCustomization.AllowSort = false;
			gridView1.OptionsView.ShowGroupPanel = false;
		}

		const string OrderFieldName = "Order";

		private void button1_Click(object sender, System.EventArgs e) {
			DevExpress.XtraGrid.Views.Grid.GridView view = gridView1;
			view.GridControl.Focus();
			int index = view.FocusedRowHandle;
			if(index <= 0) return;
			
			DataRow row1 = view.GetDataRow(index);
			DataRow row2 = view.GetDataRow(index - 1);
			object val1 = row1[OrderFieldName];
			object val2 = row2[OrderFieldName];
			row1[OrderFieldName] = val2;
			row2[OrderFieldName] = val1;

			view.FocusedRowHandle = index - 1;
		}

		private void button2_Click(object sender, System.EventArgs e) {
			DevExpress.XtraGrid.Views.Grid.GridView view = gridView1;
			view.GridControl.Focus();
			int index = view.FocusedRowHandle;
			if(index >= view.DataRowCount - 1) return;

			DataRow row1 = view.GetDataRow(index);
			DataRow row2 = view.GetDataRow(index + 1);
			object val1 = row1[OrderFieldName];
			object val2 = row2[OrderFieldName];
			row1[OrderFieldName] = val2;
			row2[OrderFieldName] = val1;

			view.FocusedRowHandle = index + 1;
		}
	}
}

File: VB/Form1.vb
Imports System
Imports System.Drawing
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data

Namespace SwapRows

    ''' <summary>
    ''' Summary description for Form1.
    ''' </summary>
    Public Class Form1
        Inherits Form

        Private gridControl1 As DevExpress.XtraGrid.GridControl

        Private gridView1 As DevExpress.XtraGrid.Views.Grid.GridView

        Private dataSet1 As DataSet

        Private dataTable1 As DataTable

        Private dataColumn1 As DataColumn

        Private dataColumn2 As DataColumn

        Private dataColumn3 As DataColumn

        Private button1 As Button

        Private button2 As Button

        ''' <summary>
        ''' Required designer variable.
        ''' </summary>
        Private components As System.ComponentModel.Container = Nothing

        Public Sub New()
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()
        '
        ' TODO: Add any constructor code after InitializeComponent call
        '
        End Sub

        ''' <summary>
        ''' Clean up any resources being used.
        ''' </summary>
        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If components IsNot Nothing Then
                    components.Dispose()
                End If
            End If

            MyBase.Dispose(disposing)
        End Sub

#Region "Windows Form Designer generated code"
        ''' <summary>
        ''' Required method for Designer support - do not modify
        ''' the contents of this method with the code editor.
        ''' </summary>
        Private Sub InitializeComponent()
            gridControl1 = New DevExpress.XtraGrid.GridControl()
            gridView1 = New DevExpress.XtraGrid.Views.Grid.GridView()
            dataSet1 = New DataSet()
            dataTable1 = New DataTable()
            dataColumn1 = New DataColumn()
            dataColumn2 = New DataColumn()
            dataColumn3 = New DataColumn()
            button1 = New Button()
            button2 = New Button()
            CType(gridControl1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(gridView1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(dataSet1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(dataTable1, System.ComponentModel.ISupportInitialize).BeginInit()
            Me.SuspendLayout()
            ' 
            ' gridControl1
            ' 
            gridControl1.Dock = DockStyle.Left
            ' 
            ' gridControl1.EmbeddedNavigator
            ' 
            gridControl1.EmbeddedNavigator.Name = ""
            gridControl1.Location = New System.Drawing.Point(0, 0)
            gridControl1.MainView = gridView1
            gridControl1.Name = "gridControl1"
            gridControl1.Size = New System.Drawing.Size(267, 235)
            gridControl1.TabIndex = 0
            gridControl1.ViewCollection.AddRange(New DevExpress.XtraGrid.Views.Base.BaseView() {gridView1})
            ' 
            ' gridView1
            ' 
            gridView1.GridControl = gridControl1
            gridView1.Name = "gridView1"
            ' 
            ' dataSet1
            ' 
            dataSet1.DataSetName = "NewDataSet"
            dataSet1.Locale = New Globalization.CultureInfo("en-US")
            dataSet1.Tables.AddRange(New DataTable() {dataTable1})
            ' 
            ' dataTable1
            ' 
            dataTable1.Columns.AddRange(New DataColumn() {dataColumn1, dataColumn2, dataColumn3})
            dataTable1.TableName = "Table1"
            ' 
            ' dataColumn1
            ' 
            dataColumn1.ColumnName = "ID"
            dataColumn1.DataType = GetType(Integer)
            ' 
            ' dataColumn2
            ' 
            dataColumn2.ColumnName = "Keyword"
            ' 
            ' dataColumn3
            ' 
            dataColumn3.ColumnName = "Order"
            dataColumn3.DataType = GetType(Integer)
            ' 
            ' button1
            ' 
            button1.Location = New System.Drawing.Point(280, 28)
            button1.Name = "button1"
            button1.Size = New System.Drawing.Size(62, 20)
            button1.TabIndex = 1
            button1.Text = "&Up"
            AddHandler button1.Click, New EventHandler(AddressOf button1_Click)
            ' 
            ' button2
            ' 
            button2.Location = New System.Drawing.Point(280, 62)
            button2.Name = "button2"
            button2.Size = New System.Drawing.Size(62, 20)
            button2.TabIndex = 2
            button2.Text = "&Down"
            AddHandler button2.Click, New EventHandler(AddressOf button2_Click)
            ' 
            ' Form1
            ' 
            AutoScaleBaseSize = New System.Drawing.Size(5, 13)
            ClientSize = New System.Drawing.Size(353, 235)
            Me.Controls.Add(button2)
            Me.Controls.Add(button1)
            Me.Controls.Add(gridControl1)
            Name = "Form1"
            Text = "Form1"
            AddHandler Load, New EventHandler(AddressOf Form1_Load)
            CType(gridControl1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(gridView1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(dataSet1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(dataTable1, System.ComponentModel.ISupportInitialize).EndInit()
            Me.ResumeLayout(False)
        End Sub

#End Region
        ''' <summary>
        ''' The main entry point for the application.
        ''' </summary>
        <STAThread>
        Shared Sub Main()
            Call Application.Run(New Form1())
        End Sub

        Private Sub FillDataSet()
            dataTable1.Rows.Add(New Object() {1, "abstract", 1})
            dataTable1.Rows.Add(New Object() {2, "event", 2})
            dataTable1.Rows.Add(New Object() {3, "new", 3})
            dataTable1.Rows.Add(New Object() {4, "struct", 4})
            dataTable1.Rows.Add(New Object() {5, "as", 5})
            dataTable1.Rows.Add(New Object() {6, "explicit", 6})
            dataTable1.Rows.Add(New Object() {7, "null", 7})
            dataTable1.Rows.Add(New Object() {8, "switch", 8})
            dataTable1.Rows.Add(New Object() {9, "base", 9})
            dataTable1.Rows.Add(New Object() {10, "extern", 10})
            dataTable1.Rows.Add(New Object() {11, "object", 11})
            dataTable1.Rows.Add(New Object() {12, "this", 12})
            dataTable1.Rows.Add(New Object() {13, "bool", 13})
            dataTable1.Rows.Add(New Object() {14, "false", 14})
            dataTable1.Rows.Add(New Object() {15, "operator", 15})
            dataTable1.Rows.Add(New Object() {16, "throw", 16})
            dataTable1.Rows.Add(New Object() {17, "break", 17})
            dataTable1.AcceptChanges()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            FillDataSet()
            gridControl1.DataSource = dataTable1
            gridView1.PopulateColumns()
            gridView1.Columns(OrderFieldName).SortOrder = DevExpress.Data.ColumnSortOrder.Ascending
            gridView1.OptionsCustomization.AllowSort = False
            gridView1.OptionsView.ShowGroupPanel = False
        End Sub

        Const OrderFieldName As String = "Order"

        Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim view As DevExpress.XtraGrid.Views.Grid.GridView = gridView1
            view.GridControl.Focus()
            Dim index As Integer = view.FocusedRowHandle
            If index <= 0 Then Return
            Dim row1 As DataRow = view.GetDataRow(index)
            Dim row2 As DataRow = view.GetDataRow(index - 1)
            Dim val1 As Object = row1(OrderFieldName)
            Dim val2 As Object = row2(OrderFieldName)
            row1(OrderFieldName) = val2
            row2(OrderFieldName) = val1
            view.FocusedRowHandle = index - 1
        End Sub

        Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim view As DevExpress.XtraGrid.Views.Grid.GridView = gridView1
            view.GridControl.Focus()
            Dim index As Integer = view.FocusedRowHandle
            If index >= view.DataRowCount - 1 Then Return
            Dim row1 As DataRow = view.GetDataRow(index)
            Dim row2 As DataRow = view.GetDataRow(index + 1)
            Dim val1 As Object = row1(OrderFieldName)
            Dim val2 As Object = row2(OrderFieldName)
            row1(OrderFieldName) = val2
            row2(OrderFieldName) = val1
            view.FocusedRowHandle = index + 1
        End Sub
    End Class
End Namespace
